src/server.c

```diff
@@ -4291,72 +4323,75 @@ void addReplyCommand(client *c, struct redisCommand *cmd) {
         }
 
-        /* We are adding: command name, arg count, flags, first, last, offset, categories, additional information (map) */
-        addReplyArrayLen(c, 8);
-        addReplyBulkCString(c, cmd->name);
+        addReplyArrayLen(c, !cmd->subcommands_dict? 9: 10);
+        if (show_full_name && cmd->parent) {
+            sds fullname = sdscatfmt(sdsempty(), "%s %s", cmd->parent->name, cmd->name);
+            addReplyBulkSds(c, fullname);
+        } else
+            addReplyBulkCString(c, cmd->name);
         addReplyLongLong(c, cmd->arity);
         addReplyFlagsForCommand(c, cmd);
         addReplyLongLong(c, firstkey);
         addReplyLongLong(c, lastkey);
         addReplyLongLong(c, keystep);
         addReplyCommandCategories(c, cmd);
-        long maplen = 0;
-        void *mapreply = addReplyDeferredLen(c);
-        addReplyBulkCString(c, "summary");
-        addReplyBulkCString(c, cmd->summary);
-        maplen++;
-        addReplyBulkCString(c, "since");
-        addReplyBulkCString(c, cmd->since);
-        maplen++;
-        addReplyBulkCString(c, "group");
-        addReplyBulkCString(c, COMMAND_GROUP_STR[cmd->group]);
-        maplen++;
-        if (cmd->complexity) {
-            addReplyBulkCString(c, "complexity");
-            addReplyBulkCString(c, cmd->complexity);
-            maplen++;
-        }
-        if (cmd->doc_flags) {
-            addReplyBulkCString(c, "doc_flags");
-            addReplyDocFlagsForCommand(c, cmd);
-            maplen++;
-        }
-        if (cmd->deprecated_since) {
-            addReplyBulkCString(c, "deprecated_since");
-            addReplyBulkCString(c, cmd->deprecated_since);
-            maplen++;
-        }
-        if (cmd->replaced_by) {
-            addReplyBulkCString(c, "replaced_by");
-            addReplyBulkCString(c, cmd->replaced_by);
-            maplen++;
-        }
-        if (cmd->history) {
-            addReplyBulkCString(c, "history");
-            addReplyCommandHistory(c, cmd);
-            maplen++;
-        }
-        if (cmd->hints) {
-            addReplyBulkCString(c, "hints");
-            addReplyCommandHints(c, cmd);
-            maplen++;
-        }
-        if (cmd->args) {
-            addReplyBulkCString(c, "arguments");
-            addReplyCommandArgList(c, cmd->args);
-            maplen++;
-        }
-        if (cmd->key_specs_num) {
-            addReplyBulkCString(c, "key_specs");
-            addReplyCommandKeySpecs(c, cmd);
-            maplen++;
-        }
-        if (cmd->subcommands_dict) {
-            addReplyBulkCString(c, "subcommands");
-            addReplyCommandSubCommands(c, cmd);
-            maplen++;
-        }
-        setDeferredMapLen(c, mapreply, maplen);
+        addReplyCommandHints(c, cmd);
+        addReplyCommandKeySpecs(c, cmd);
+        if (cmd->subcommands_dict) /* don't add sub-commands on sub-commands */
+            addReplyCommandSubCommands(c, cmd, addReplyCommandInfo, 0);
```

Comment:
this is dangerous - what if we add a new element in the next version? the indices here must be constant
we need to reply with an empty array instead

Reply:
yes, i thought about it too.. so if we add another element we'll have to remove that `if`.
but do you think there's a risk keeping this code for now? i.e. that for some reason clients will write code that is not compatible with removal of the `if`?

Reply:
imagine a client that tests the length of the array reply in order to know if this command has subcommands

Reply:
ok. i'm imagining (i have a word for it, but i'll keep it to myself).
now imagine a client that fails on assertion if there are more than 7 elements in the response of COMMAND....

but anyway, if you think it's better, i don't mind adding it (was just thinking of saving some extra bytes)..
do you want to put an empty array, or a nil?

Reply:
adding an empty array.